xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -264,7 +284,85 @@ protected void handleLdsResponse(String versionInfo, List<Any> resources, String
       ResourceSubscriber subscriber = ldsResourceSubscribers.get(resource);
       if (ldsUpdates.containsKey(resource)) {
-        subscriber.onData(ldsUpdates.get(resource));
+        subscriber.onData(ldsUpdates.get(resource), null, null, 0);
+      } else {
+        subscriber.onAbsent();
+      }
+    }
+    for (String resource : rdsResourceSubscribers.keySet()) {
+      if (!rdsNames.contains(resource)) {
+        ResourceSubscriber subscriber = rdsResourceSubscribers.get(resource);
+        subscriber.onAbsent();
+      }
+    }
+  }
+
+  @Override
+  protected void handleLdsResponse(String versionInfo, List<Any> resources, String nonce) {
+    Map<String, LdsUpdate> ldsUpdates = new HashMap<>(resources.size());
+    Map<String, Any> listenerProtos = new HashMap<>(resources.size());
+    Set<String> listenerNames = new HashSet<>(resources.size());
+    List<String> errors = new ArrayList<>();
+    Set<String> rdsNames = new HashSet<>();
+    long updateTime = timeProvider.currentTimeNanos();
+
+    for (int i = 0; i < resources.size(); i++) {
+      // Unpack the Listener.
+      Any rawListener = resources.get(i);
+      Listener listener;
+      try {
+        listener = unpackCompatibleType(rawListener, Listener.class, ResourceType.LDS.typeUrl(),
+            ResourceType.LDS.typeUrlV2());
+      } catch (InvalidProtocolBufferException e) {
+        errors.add("LDS response Resource index " + i + " - can't decode Listener: " + e);
+        continue;
+      }
+      String listenerName = listener.getName();
+      listenerNames.add(listenerName);
+
+      // Process Listener into LdsUpdate.
+      LdsUpdate ldsUpdate;
+      try {
+        if (listener.hasApiListener()) {
+          ldsUpdate = processClientSideListener(listener);
+        } else {
+          ldsUpdate = processServerSideListener(listener);
+        }
+      } catch (ResourceInvalidException e) {
+        errors.add("LDS response Listener '" + listenerName + "' validation error: " + e);
+        continue;
+      }
+
+      // LdsUpdate parsed successfully.
+      ldsUpdates.put(listenerName, ldsUpdate);
+      listenerProtos.put(listenerName, rawListener);
+      if (ldsUpdate.rdsName != null) {
+        rdsNames.add(ldsUpdate.rdsName);
+      }
+    }
+    getLogger().log(XdsLogLevel.INFO,
+        "Received LDS Response version {0} nonce {1}. Parsed resources: {2}",
+        versionInfo, nonce, listenerNames);
+
+    // NACK on errors.
+    if (!errors.isEmpty()) {
+      String errorDetail = combineErrors(errors);
+      getLogger().log(XdsLogLevel.WARNING,
+          "Failed processing LDS Response version {0} nonce {1}. Errors:\n{2}",
+          versionInfo, nonce, errorDetail);
+      attachErrorStateToMetadata(
+          ResourceType.LDS, listenerNames, versionInfo, updateTime, errorDetail);
+      nackResponse(ResourceType.LDS, nonce, errorDetail);
+      return;
+    }
+
+    // ACK on success.
+    ackResponse(ResourceType.LDS, versionInfo, nonce);
```

Comment:
1) `handleResourcesAcked`—I was thinking the same thing, but wasn't sure about this because I noticed the difference in behavior of handleRdsResponse, handleEdsResponse—we don't call onAbsent() as in LDS, CDS. Is there a good way to reconcile this?
2) Rename `attachErrorStateToMetadata` to `handleResourcesNacked`: good idea!
3) Group parsed/raw pair—was considering this as well, but abandoned this idea because I wasn't sure what's a good way to do this in Java. Should this be a `Map<String, aDataStructureToHoldBothThings>`?

Reply:
> I noticed the difference in behavior of handleRdsResponse, handleEdsResponse—we don't call onAbsent() as in LDS, CDS. Is there a good way to reconcile this?

Yes, they are different for LDS/CDS and RDS/EDS. RDS and EDS are incremental protocol, receiving the response not containing some resource doesn't imply the resource does not exist. However, LDS and CDS are state-of-the-world, each response should include all resources of that type. I guess you could just do a if based on the resource type in `handleResourceAcked()`.